### ISSUE TITLE
`rustup update` should run self update first

### ISSUE THREAD

**Issue by killercup** [rust-lang-nursery/rustup.rs#500]
Currently, running `rustup update` updates all toolchains _and then_ runs `self update`.

It might be wise to run `self update` first, to pick up

- possible changes in the metadata format,
- security fixes for downloading,
- new compression options,
- and other improvements

as soon as possible and use them for updating the toolchain.

One difficulty might be that rustup will need to restart itself/start the new version when there was an update before running the toolchain updates.

**Comment by brson** [comment]
I agree that it would be better to do the self-update first, but it's very tricky. On windows, we can't replace the rustup executable while rustup is running, so the last thing `rustup update` does is *exit successfully*, then a separate process overwrites rustup.exe. So there's a chance the update actually fails but doesn't get reported. If we were to do incompatible metadata changes before overwriting the exe, and then the self-update fails, rustup could be left completely broken.

**Comment by Diggsey** [comment]
You can *rename* a running process though, so we can have rustup download the new version, rename itself to a temp directory, call the new rustup with the original update command line (piping stdin/stdout through the old rustup), and then delete itself. If the self deletion fails we get a left-over executable in your temp directory rather than a broken installation.

**Comment by brson** [comment]
Ah yes that does seem doable.

**Comment by vks** [comment]
Not running `rustup self update` first seems to have caused upgrades to fail for a lot of people.
